File: Assignment 4/task3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "t3.h"

int visited[MAX_NB];
int adjMat[MAX_NB][MAX_NB];

int next_field( FILE *f, char *buf, int max ) { //from assignment 0
    int i=0, end=0, quoted=0;
    
    for(;;) {
        // fetch the next character from file
        buf[i] = fgetc(f);
        // if we encounter quotes then flip our state and immediately fetch next char
        if(buf[i]=='"') { quoted=!quoted; buf[i] = fgetc(f); }
        // end of field on comma if we're not inside quotes
        if(buf[i]==',' && !quoted) { break; }
        // end record on newline or end of file
        if(feof(f) || buf[i]=='\n') { end=1; break; }
        // truncate fields that would overflow the buffer
        if( i<max-1 ) { ++i; }
    }

    buf[i] = 0;
    return end;
}

void getStops (  FILE *csv, struct stop *p) {
    char buf[MAX_BUFFER];
    
    next_field( csv, buf, MAX_BUFFER );
    p->id = atoi(buf);
    next_field( csv, p->title, MAX_BUFFER );
    next_field( csv, buf, MAX_BUFFER );
    p->lat = atof(buf);
    next_field( csv, buf, MAX_BUFFER );
    p->lont = atof(buf);
}

Graph* create_graph(int num_nodes){
    Graph* g = (Graph *)malloc(sizeof(struct Graph));
    g->vertices=num_nodes;

    int i,j;
    for(i=0;i<num_nodes;i++){
        for(j=0;j<num_nodes;j++){
            adjMat[i][j]=0;
        }
    }
    int n=num_nodes;
    return g;
}
Graph *graph;

int load_vertices(char *fname){
    FILE *f;
    struct stop pArray[MAX_NB];
    struct stop p;
    f=fopen(fname,"r");
    if(!f) {
        printf("unable to open file\n");
        return 0;
    }
    getStops( f, &p );
    int ngames = 0;
    int fr;
    while(!feof(f)) {
        getStops( f, &pArray[ngames]);
        fr=pArray[ngames].id;
        arr[fr]=&pArray[ngames];
        ngames++;
    }
    printf("Loaded %d vertices\n",ngames);
    fclose(f);
    graph = create_graph(ngames);
    
    return 1;
}


void getEdges (  FILE *csv, struct edges *p) {
    char buf[MAX_BUFFER];
    next_field( csv, buf, MAX_BUFFER );
    p->from = atoi(buf);
    next_field( csv, buf, MAX_BUFFER );
    p->to = atoi(buf);
    next_field( csv, buf, MAX_BUFFER );
    p->weight = atoi(buf);
}

void add_edge(int from, int to, int weight){
    adjMat[from][to]=weight;
    adjMat[to][from]=weight;
}

int load_edges(char *fname){
    FILE *f;
    struct edges pArray[MAX_NB];
    struct edges p;
    f=fopen(fname,"r");
    if(!f) {
        printf("unable to open file\n");
        return 0;
    }
    getEdges( f, &p );
    int nedges = 0;
    while(!feof(f)) {
    getEdges( f, &pArray[nedges]);
    nedges++;
    }
    int i;
    for(i=0;i<nedges;i++){
    add_edge(pArray[i].from,pArray[i].to,pArray[i].weight);
    }
    printf("Loaded %d edges\n",nedges);
    fclose(f);
    return 1;
}

void pathPrint(int p[], int q){
    if (p[q] == - 1)
        return;
    pathPrint(p, p[q]);
    printf("%d %s\n",arr[q]->id, arr[q]->title);
}

void shortest_path(int startNode, int endNode){
    int distance[MAX_NB];
    int pre[MAX_NB];
    int count,minDistance,nextNode,i,j;
    pre[startNode]=-1;
    int n=MAX_NB;
    
    
    for (i = 0; i < n; i++){
    for (j = 0; j < n; j++){
      if (adjMat[i][j] == 0)
        adjMat[i][j] = INFINITY;
        }
    }

    for (i = 0; i <n; i++) {
    distance[i] = INFINITY;
  }

    distance[startNode] = 0;
    count = 1;

    while (count < n-1) {
        minDistance = INFINITY;
        for (i = 0; i < n; i++){
              if ((distance[i] < minDistance) && (visited[i])!=1) {
                minDistance = distance[i];
                nextNode = i;
              }}
        
        visited[nextNode] = 1;
    
        for (i = 0; i < n; i++)
              if (!(visited[i]))
                if (minDistance + adjMat[nextNode][i] < distance[i]) {
                  distance[i] = minDistance + adjMat[nextNode][i];
                  pre[i]=nextNode;
                }
        count++;
    }
    printf("%d %s\n",arr[startNode]->id, arr[startNode]->title);
    pathPrint(pre,endNode);
}

void delete_graph(Graph* g){
    free(g);
}

void free_memory(void){
    delete_graph(graph);
}

File: Assignment 2/task3.py
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX 50

struct t4_ign
{
    char title[MAX];
    char platform[MAX];
    int score;
    int release_year;
    struct t4_ign* following;
}*head = NULL, *current, *newcur;

// The CSV parser
int next_field( FILE *f, char *buf, int max ) {
    int i=0, end=0, quoted=0;
    
    for(;;) {
        // fetch the next character from file
        buf[i] = fgetc(f);
        // if we encounter quotes then flip our state and immediately fetch next char
        if(buf[i]=='"') { quoted=!quoted; buf[i] = fgetc(f); }
        // end of field on comma if we're not inside quotes
        if(buf[i]==',' && !quoted) { break; }
        // end record on newline or end of file
        if(feof(f) || buf[i]=='\n') { end=1; break; }
        // truncate fields that would overflow the buffer
        if( i<max-1 ) { ++i; }
    }

    buf[i] = 0; // null terminate the string
    return end; // flag stating whether or not this is end of the line
}

// From assignment 0
void fetch_file (  FILE *csv, struct t4_ign *point) {
    char buf[MAX];
    next_field( csv, point->title, MAX );      // name and type are just strings so read
    next_field( csv, point->platform, MAX );   // those directly into the struct
    next_field( csv, buf, MAX ); // atoi stands for ASCII to Integer
    point->score = atoi(buf);     // It converts strings to numbers
    next_field( csv, buf, MAX ); // It is not a totally safe function to use.
    point->release_year = atoi(buf);
    
}

void insertValue(struct t4_ign** order, struct t4_ign* cur){
    if(*order==NULL || (*order)->score <= cur->score){
        cur->following = *order;
        *order = cur;
    }
    else{
        newcur = *order;
        while(newcur->following!=NULL && newcur->score >= cur->score)
            newcur=newcur->following;
        cur->following=newcur->following;
        newcur->following=cur;
    }
}

void sort(struct t4_ign **head)
{
    struct t4_ign *sort = NULL;
    current = *head;
    while (current != NULL)
    {
        struct t4_ign *p = current->following;
        insertValue(&sort, current);
        current = p;
    }
    *head = sort;
}

void list(FILE *f)
{
    while(!feof(f))
         {
             struct t4_ign *new;
             new=(struct t4_ign *)malloc(sizeof(struct t4_ign));
             fetch_file( f, new );
             new->following=NULL;
             if(head==NULL){
                 head=new;
                 current=new;
             }
             else{
                 current->following=new;
                 current=new;
             }
         }
}

void printList(struct t4_ign *p )
{
    for (int i=0; i<10;i++)
    {
        p=p->following;
        printf("\n %s\t\t%s\t\t%d\t\t%d\n",p->title,p->platform,p->score,p->release_year);
    }
}


int main ( int argc, char *argv[] ) {
    FILE *f = fopen(argv[1], "r");
    if(!f) {
        printf("unable to open %s\n", argv[1]);
        return EXIT_FAILURE;
    }
    
    list(f);
    sort(&head);
    printList(head);
    fclose(f);
    return EXIT_SUCCESS;
}
